test/factories/smart_proxy.rb

```diff
@@ -25,93 +33,20 @@
       end
     end
 
-    factory :template_smart_proxy do
-      after(:build) do |smart_proxy, _evaluator|
-        smart_proxy.smart_proxy_features << FactoryBot.build(:smart_proxy_feature, :templates, :smart_proxy => smart_proxy)
-      end
-    end
-
-    factory :bmc_smart_proxy do
-      after(:build) do |smart_proxy, _evaluator|
-        smart_proxy.smart_proxy_features << FactoryBot.build(:smart_proxy_feature, :bmc, :smart_proxy => smart_proxy)
-      end
-    end
-
-    factory :dhcp_smart_proxy do
-      after(:build) do |smart_proxy, _evaluator|
-        smart_proxy.smart_proxy_features << FactoryBot.build(:smart_proxy_feature, :dhcp, :smart_proxy => smart_proxy, :capabilities => ["dhcp_filename_ipv4"])
-      end
-    end
-
-    factory :dns_smart_proxy do
-      after(:build) do |smart_proxy, _evaluator|
-        smart_proxy.smart_proxy_features << FactoryBot.build(:smart_proxy_feature, :dns, :smart_proxy => smart_proxy)
-      end
-    end
-
-    factory :ipam_smart_proxy do
-      after(:build) do |smart_proxy, _evaluator|
-        smart_proxy.smart_proxy_features << FactoryBot.build(:smart_proxy_feature, :external_ipam, :smart_proxy => smart_proxy)
-      end
-    end
-
-    factory :httpboot_smart_proxy do
-      after(:build) do |smart_proxy, _evaluator|
-        smart_proxy.smart_proxy_features << FactoryBot.build(:smart_proxy_feature, :httpboot, :smart_proxy => smart_proxy)
-      end
-    end
-
-    factory :puppet_ca_smart_proxy do
-      before(:create, :build, :build_stubbed) do
-        ProxyAPI::V2::Features.any_instance.stubs(:features).returns(:puppetca => {'state' => 'running'})
-      end
-      after(:build) do |smart_proxy, _evaluator|
-        smart_proxy.smart_proxy_features << FactoryBot.build(:smart_proxy_feature, :puppetca, :smart_proxy => smart_proxy)
-      end
-    end
-
-    factory :realm_smart_proxy do
-      after(:build) do |smart_proxy, _evaluator|
-        smart_proxy.smart_proxy_features << FactoryBot.build(:smart_proxy_feature, :realm, :smart_proxy => smart_proxy)
+    proxy_features.each do |feature|
+      trait feature do
+        smart_proxy_features do
+          [association(:smart_proxy_feature, feature, :smart_proxy => @instance)]
```

Comment:
https://thoughtbot.github.io/factory_bot/cookbook/interconnected-associations.html said you should be able to use `instance`, but I only got it to work with `@instance`.